src/classes/BaseHttpUrlEncodedClient.cls

```
@@ -0,0 +1,98 @@
+      * @param body: Request Body. Can only be map
+      * @return Void.  Mutates req.body
+    */
+    protected virtual override void setRequestBody(HttpRequest req, Object body){
+        if(body != null){
+            if(body instanceOf Map<String,String>){
+                req.setBody(buildEntityBody((Map<string, string>) body));
+            }else{
+                super.setRequestBody(req, body);
+            }
+        }
+    }
+
+
+    protected virtual override HttpRequest genRequest(String method, String uri){
+        HttpRequest req = super.genRequest(method,uri);
+        req.setHeader('Content-Type', 'application/x-www-form-urlencoded');
+        return req;
+    }
+
+    /* parse JSON/XML/Other to response DTO type */
+    protected virtual Object parseResponse(HttpResponse resp, Type returnType){
+        String resBody = resp.getBody();
+        return JSON.deserialize(resBody, returnType);
+    }
+
+
+    /* parse json exception */
+    public class ResponseJsonParseException extends Exception {
+        public HttpResponse resp;
+        public Type returnType;
+
+        public ResponseJsonParseException(HttpResponse resp, Type returnType){
+            this.resp = resp;
+            this.returnType = returnType;
+            this.setMessage('Failed to parse http response to type: ' + returnType.toString());
+        }
+    }
+
+    /**
+     * Builds the entity body. and URL encodes it
+     *
+     * @param params
+     *            the params
+     * @return the url encoded form entity
+     */
+    private String buildEntityBody(Map<String,String> params) {
+        String entityBody='';
+        if (params != null) {
+            for (String key : params.keySet()) {
+                try {
+                    entityBody += (entityBody=='' ? '' : '&')
+                                    + key + '=' + EncodingUtil.urlEncode(params.get(key), 'UTF-8');
+                } catch (Exception e) {
+                    System.debug('ERROR: ' + e);
```

Probably don't want to just eat the exception.  Should just throw